cellbase-lib/src/main/java/org/opencb/cellbase/lib/impl/core/singleton/DataReleaseSingleton.java

```diff
@@ -0,0 +1,218 @@
+                                    Arrays.toString(e.getStackTrace()));
+                            try {
+                                // Sleep 5 sec before retrying to avoid tight loop
+                                Thread.sleep(5000);
+                            } catch (InterruptedException ie) {
+                                // Restore interrupt status
+                                Thread.currentThread().interrupt();
+                                break; // Exit loop if interrupted
+                            }
+                        }
+                    }
+                }).start();
+            }
+        }
+    }
+
+    // Initialization method to set up the instance with parameters
+    public static synchronized void initialize(CellBaseManagerFactory managerFactory) throws CellBaseException {
+        if (instance == null) {
+            instance = new DataReleaseSingleton(managerFactory);
+        }
+    }
+
+    // Method to get the single instance of the class
+    public static DataReleaseSingleton getInstance() {
+        if (instance == null) {
+            throw new IllegalStateException("Singleton not initialized. Call the function 'initialize' first.");
+        }
+        return instance;
+    }
+
+    // Method to load data from MongoDB and cache it
+    private void loadData(String dbname) throws CellBaseException {
+        DatabaseInfo dbInfo = getDatabaseInfo(dbname);
+
+        String species = dbInfo.getSpecies();
+        String assembly = dbInfo.getAssembly();
+        ReleaseMongoDBAdaptor releaseMongoDBAdaptor = managerFactory.getDataReleaseManager(species, assembly).getReleaseDBAdaptor();
+        List<DataRelease> dataReleases = releaseMongoDBAdaptor.getAll().getResults();
+        if (CollectionUtils.isNotEmpty(dataReleases)) {
+            for (DataRelease dataRelease : dataReleases) {
+                Map<String, MongoDBCollection> collectionMap = new HashMap<>();
+                for (Map.Entry<String, String> entry : dataRelease.getCollections().entrySet()) {
+                    collectionMap.put(entry.getKey(), releaseMongoDBAdaptor.getMongoDataStore().getCollection(entry.getValue()));
+                }
+                dbInfo.getCacheData().put(dataRelease.getRelease(), collectionMap);
+            }
+        }
+    }
+
+    public void checkDataRelease(String species, String assembly, String version, int release) throws CellBaseException {
+        checkDataRelease(MongoDBManager.getDatabaseName(species, assembly, version), release);
+    }
+
+    public void checkDataRelease(String dbname, int release) throws CellBaseException {
+        checkDataRelease(dbname, release, null);
+    }
+
+    public void checkDataRelease(String dbname, int release, String data) throws CellBaseException {
+        DatabaseInfo dbInfo = getDatabaseInfo(dbname);
+
+        // Lock and load data if necessary
+        dbInfo.getRwLock().writeLock().lock();
```

Comment:
This would be an unneeded lock if `dbInfo.getCacheData().containsKey(release)` and `dbInfo.getCacheData().get(release).containsKey(data)`. Given that this method will be called a lot of times, might be worth improving the logic?